hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/inject/AvroUpsertTarget.java

```diff
@@ -40,60 +65,59 @@ class AvroUpsertTarget implements UpsertTarget {
     }
 
     @Override
+    @SuppressWarnings("ReturnCount")
     public UpsertInjector createInjector(@Nullable String path, QueryDataType type) {
         if (path == null) {
             return FAILING_TOP_LEVEL_INJECTOR;
         }
 
-        switch (type.getTypeFamily()) {
-            case TINYINT:
-                return value -> record.set(path, value == null ? null : ((Byte) value).intValue());
-            case SMALLINT:
-                return value -> record.set(path, value == null ? null : ((Short) value).intValue());
+        Schema fieldSchema = schema.getField(path).schema();
+        Schema.Type schemaFieldType = unwrapNullableType(fieldSchema).getType();
+        switch (schemaFieldType) {
             case BOOLEAN:
-            case INTEGER:
-            case BIGINT:
-            case REAL:
+            case INT:
+            case LONG:
+            case FLOAT:
             case DOUBLE:
-                return value -> record.set(path, value);
-            case DECIMAL:
-            case TIME:
-            case DATE:
-            case TIMESTAMP:
-            case TIMESTAMP_WITH_TIME_ZONE:
-            case VARCHAR:
-                return value -> record.set(path, QueryDataType.VARCHAR.convert(value));
-            case OBJECT:
-                return createObjectInjector(path);
+            case STRING:
+                return value -> {
+                    try {
+                        record.set(path, AVRO_TO_SQL.get(schemaFieldType).convert(value));
+                    } catch (QueryException e) {
+                        throw QueryException.error("Cannot convert " + value + " to " + schemaFieldType
+                                + " (field=" + path + ")");
+                    }
+                };
+            case UNION:
+                return value -> {
+                    if (value == null) {
+                        record.set(path, null);
+                        return;
+                    } else {
+                        for (Schema.Type target : CONVERSION_PREFS.get(value.getClass())) {
+                            if (fieldSchema.getTypes().stream().anyMatch(schema -> schema.getType() == target)) {
+                                try {
+                                    record.set(path, AVRO_TO_SQL.get(target).convert(value));
+                                    return;
+                                } catch (QueryException ignored) { }
```

Comment:
why the exception is ignored?

Reply:
Available conversion targets are tried in the preferred order until one of them can successfully convert the value. A conversion may fail due to numeric overflow, not-a-number, infinity, etc.

Reply:
what happens if all conversions fail? Do you assume that there will always be toString() conversion which is not supposed to fail? If so, it is worth documenting and this is highly non-trivial. Also STRING is not always the last entry in `CONVERSION_PREFS`.

Reply:
ah, there is `return` inside `try`. IMO still worth adding a comment

Reply:
another related comment about strings: https://github.com/hazelcast/hazelcast/pull/25207#discussion_r1300708734

We can merge this as-is and discuss various approaches to union matching in separate issue/PR.

Reply:
Union types are ordered, but that order shouldn't be used because converters don't warn of precision loss. What I understand from a "union" is that we should try our best in order not to loose any information by intelligently choosing a target type. Also, it would be better if this choice remains constant for the same source type.

We have different conversion preferences for different source types because if we apply
1. (`INT`, `LONG`, `FLOAT`, `DOUBLE`, `STRING`) for all, floating-point numbers may underflow; and
2. (`FLOAT`, `DOUBLE`, `INT`, `LONG`, `STRING`) for all, integers may loose precision.

Finally, numeric literals take the narrowest —or the most logical— type possible already. If the number comes from a mapping, it can be casted. Otherwise, switching among target types depending on the precision might sound like a nice storage optimization, but it will be confusing for the consumer.

Reply:
https://hazelcast.atlassian.net/browse/HZ-3057